File: app/src/main/kotlin/com/mattdixon/jobjar/ui/components/SubtasksSection.kt
package com.mattdixon.jobjar.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.ChevronRight
import androidx.compose.material.icons.filled.Lock
import androidx.compose.material.icons.filled.RadioButtonUnchecked
import androidx.compose.material3.Card
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import com.mattdixon.jobjar.R
import com.mattdixon.jobjar.data.Job
import com.mattdixon.jobjar.data.JobRepository
import com.mattdixon.jobjar.data.LONG_JOB_MINUTES
import com.mattdixon.jobjar.data.isUnblocked
import com.mattdixon.jobjar.data.remainingMinutesOf
import com.mattdixon.jobjar.data.unallocatedMinutesOf
import com.mattdixon.jobjar.ui.theme.Spacing
import com.mattdixon.jobjar.util.formatMinutes
import kotlinx.coroutines.launch

/**
 * The "Subtasks" list, progress summary, and add-subtask button for a top-level job. Shared by
 * the job detail screen and the job creation/edit screen so subtask management looks and works
 * identically everywhere it appears, instead of each screen reimplementing it.
 *
 * [parentId] must refer to a job that is itself NOT a subtask (subtasks are one level deep) -
 * callers are responsible for that check, since this composable has no way to verify it.
 */
@Composable
fun SubtasksSection(
    repository: JobRepository,
    parentId: Long,
    parentEstimatedMinutes: Int,
    onOpenSubtask: (Long) -> Unit,
    onAddSubtask: () -> Unit,
    modifier: Modifier = Modifier,
    /**
     * The "remaining of total" line (unfinished work left) is only meaningful once a job's
     * actually being worked - a freshly created subtask is never done, so it'd just sit at the
     * full total the whole time a list is being built. The add/edit screen leaves this false and
     * shows its own live [AllocationSummary] next to the duration picker instead (for both the
     * job being edited and, while adding a subtask, the parent it belongs to) - not here, so the
     * two don't compete for attention on the same screen. The detail screen leaves this true.
     */
    showRemainingSummary: Boolean = true
) {
    val subtasksFlow = remember(repository, parentId) { repository.subtasksOf(parentId) }
    val subtasks by subtasksFlow.collectAsState(initial = emptyList())
    val scope = rememberCoroutineScope()

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(Spacing.lg)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(stringResource(R.string.subtasks_title), style = MaterialTheme.typography.titleMedium)
            if (subtasks.isNotEmpty()) {
                Text(
                    stringResource(R.string.subtasks_done_count, subtasks.count { it.isDone }, subtasks.size),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }

        if (subtasks.isEmpty()) {
            Text(
                stringResource(R.string.subtasks_empty_hint),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        } else {
            if (showRemainingSummary) {
                val remaining = remainingMinutesOf(parentEstimatedMinutes, subtasks)
                Text(
                    stringResource(R.string.label_remaining_of_total, formatMinutes(remaining), formatMinutes(parentEstimatedMinutes)),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            val siblingsById = subtasks.associateBy { it.id }
            Column(verticalArrangement = Arrangement.spacedBy(Spacing.md)) {
                subtasks.forEach { subtask ->
                    SubtaskRow(
                        subtask = subtask,
                        waitingOnTitle = subtask.dependsOnSubtaskId
                            ?.takeUnless { subtask.isUnblocked(siblingsById) }
                            ?.let { siblingsById[it]?.title },
                        onClick = { onOpenSubtask(subtask.id) },
                        onToggleDone = { scope.launch { repository.toggleDone(subtask) } }
                    )
                }
            }
        }

        OutlinedButton(onClick = onAddSubtask, modifier = Modifier.fillMaxWidth()) {
            Icon(Icons.Filled.Add, contentDescription = null)
            Text(stringResource(R.string.cd_add_subtask))
        }
    }
}

/**
 * How much of the typed total [subtasksTotalMinutes] worth of subtasks add up to so far. A job
 * at or above [LONG_JOB_MINUTES] never had a firm target in the first place (see the "4+ hrs"
 * duration option) - "left to allocate" or "over by" would imply a ceiling that doesn't really
 * exist, so this just reports a running sum instead, no comparison to any total. Not private:
 * also used directly by the add/edit screen next to the duration picker itself, where
 * [subtasksTotalMinutes] is a live projection (every sibling subtask plus whatever's currently
 * typed for the one being edited) rather than a straight sum of already-saved subtasks.
 */
@Composable
fun AllocationSummary(estimatedMinutes: Int, subtasksTotalMinutes: Int) {
    val text = if (estimatedMinutes >= LONG_JOB_MINUTES) {
        stringResource(R.string.subtasks_allocation_running_total, formatMinutes(subtasksTotalMinutes))
    } else {
        when (val unallocated = unallocatedMinutesOf(estimatedMinutes, subtasksTotalMinutes)) {
            0 -> stringResource(R.string.subtasks_allocation_exact, formatMinutes(estimatedMinutes))
            else -> if (unallocated > 0) {
                stringResource(R.string.subtasks_allocation_under, formatMinutes(unallocated), formatMinutes(estimatedMinutes))
            } else {
                stringResource(R.string.subtasks_allocation_over, formatMinutes(-unallocated), formatMinutes(estimatedMinutes))
            }
        }
    }
    Text(text, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
}

/**
 * [waitingOnTitle], when non-null, means this subtask is soft-blocked on an unfinished sibling:
 * it's greyed out with a lock indicator and excluded from the jar's draw pool (see
 * [com.mattdixon.jobjar.data.JobRepository.drawJob]), but the checkbox stays fully live - it can
 * still be checked off by hand at any time.
 */
@Composable
private fun SubtaskRow(
    subtask: Job,
    waitingOnTitle: String?,
    onClick: () -> Unit,
    onToggleDone: () -> Unit
) {
    val blocked = waitingOnTitle != null
    val contentColor = if (blocked) MaterialTheme.colorScheme.onSurfaceVariant else MaterialTheme.colorScheme.onSurface
    Card(onClick = onClick, modifier = Modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(Spacing.md),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.md)
        ) {
            IconButton(onClick = onToggleDone) {
                Icon(
                    imageVector = if (subtask.isDone) Icons.Filled.CheckCircle else Icons.Filled.RadioButtonUnchecked,
                    contentDescription = stringResource(if (subtask.isDone) R.string.cd_mark_not_done else R.string.cd_mark_done),
                    tint = contentColor
                )
            }
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
                Text(subtask.title, style = MaterialTheme.typography.bodyLarge, color = contentColor)
                Row(horizontalArrangement = Arrangement.spacedBy(Spacing.sm), verticalAlignment = Alignment.CenterVertically) {
                    TimeBucketBadge(minutes = subtask.estimatedMinutes)
                    if (waitingOnTitle != null) {
                        Icon(
                            Icons.Filled.Lock,
                            contentDescription = null,
                            modifier = Modifier.padding(start = Spacing.xxs),
                            tint = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Text(
                            stringResource(R.string.label_waiting_on, waitingOnTitle),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }
            Icon(Icons.Filled.ChevronRight, contentDescription = null, tint = contentColor)
        }
    }
}

File: ios/JobJar/JobJar/Views/Components/SubtasksSection.swift
import SwiftData
import SwiftUI

/// The "Subtasks" list, progress summary, and add-subtask button for a top-level job. Shared
/// by the job detail screen and the job creation/edit screen so subtask management looks and
/// works identically everywhere it appears.
///
/// `parentId` must refer to a job that is itself NOT a subtask (subtasks are one level deep) -
/// callers are responsible for that check, since this view has no way to verify it.
struct SubtasksSection: View {
    let parentId: UUID
    let parentEstimatedMinutes: Int
    var onOpenSubtask: (UUID) -> Void
    var onAddSubtask: () -> Void

    @Environment(\.modelContext) private var modelContext
    @Query private var allJobs: [Job]

    private var repository: JobRepository { JobRepository(modelContext: modelContext) }
    private var subtasks: [Job] {
        allJobs.filter { $0.parentId == parentId }.sorted { $0.createdAt < $1.createdAt }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Text("Subtasks").font(.headline)
                Spacer()
                if !subtasks.isEmpty {
                    Text("\(subtasks.filter { $0.isDone }.count)/\(subtasks.count) done")
                        .font(.subheadline)
                        .foregroundStyle(.secondary)
                }
            }

            if subtasks.isEmpty {
                Text("Break this job into smaller pieces you can draw on their own.")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            } else {
                let remaining = remainingMinutesOf(estimatedMinutes: parentEstimatedMinutes, subtasks: subtasks)
                Text("\(formatMinutes(remaining)) left of \(formatMinutes(parentEstimatedMinutes)) total")
                    .font(.footnote)
                    .foregroundStyle(.secondary)

                let siblingsById = Dictionary(uniqueKeysWithValues: subtasks.map { ($0.id, $0) })
                VStack(spacing: 8) {
                    ForEach(subtasks) { subtask in
                        SubtaskRow(
                            subtask: subtask,
                            waitingOnTitle: waitingOnTitle(for: subtask, siblingsById: siblingsById),
                            onTap: { onOpenSubtask(subtask.id) },
                            onToggleDone: { repository.toggleDone(subtask, allJobs: allJobs) }
                        )
                    }
                }
            }

            Button(action: onAddSubtask) {
                Label("Add subtask", systemImage: "plus")
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 10)
            }
            .buttonStyle(.bordered)
        }
    }

    /// Non-nil (blocked) only when the dependency isn't satisfied yet - a soft block: the
    /// checkbox below stays fully live either way, this only changes the draw pool.
    private func waitingOnTitle(for subtask: Job, siblingsById: [UUID: Job]) -> String? {
        guard let prerequisiteId = subtask.dependsOnSubtaskId, !subtask.isUnblocked(siblingsById: siblingsById) else {
            return nil
        }
        return siblingsById[prerequisiteId]?.title
    }
}

private struct SubtaskRow: View {
    let subtask: Job
    let waitingOnTitle: String?
    let onTap: () -> Void
    let onToggleDone: () -> Void

    private var blocked: Bool { waitingOnTitle != nil }

    var body: some View {
        Button(action: onTap) {
            HStack(spacing: 8) {
                Button(action: onToggleDone) {
                    Image(systemName: subtask.isDone ? "checkmark.circle.fill" : "circle")
                        .foregroundStyle(subtask.isDone ? Color.accentColor : Color.secondary)
                }
                .buttonStyle(.plain)

                VStack(alignment: .leading, spacing: 4) {
                    Text(subtask.title)
                        .font(.body)
                        .foregroundStyle(blocked ? .secondary : .primary)
                    HStack(spacing: 6) {
                        TimeBadge(minutes: subtask.estimatedMinutes)
                        if let waitingOnTitle {
                            Image(systemName: "lock.fill")
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                            Text("Waiting on: \(waitingOnTitle)")
                                .font(.caption)
                                .foregroundStyle(.secondary)
                        }
                    }
                }
                Spacer()
                Image(systemName: "chevron.right")
                    .foregroundStyle(.secondary)
            }
            .padding(8)
            .background(RoundedRectangle(cornerRadius: 12).fill(Color(.secondarySystemBackground)))
        }
        .buttonStyle(.plain)
    }
}
